src/compiler/parser.ts

```diff
@@ -3506,8 +3603,14 @@ module ts {
                     return parseArrayLiteralExpression();
                 case SyntaxKind.OpenBraceToken:
                     return parseObjectLiteralExpression();
+                case SyntaxKind.AsyncKeyword:
+                    if (!lookAhead(nextTokenIsFunctionKeywordOnSameLine)) {
+                        break;
+                    }
+                    modifiers = parseModifiers();
+                    // fall-through

```

Comment:
the fallthrough seems too cute.  can you just do "return parseFunctionExpression(parseModifiers());"